File: solutions/1438.cpp
class Solution {
public:
    int longestSubarray(vector<int>& nums, int limit) {
        // max q
        deque<int> maxQ;
        // min q
        deque<int> minQ;
        
        int res = 0;
        int left = 0;
        for (int right = 0; right < nums.size(); ++right) {
            while (!maxQ.empty() && nums[right] >= nums[maxQ.back()]) maxQ.pop_back();
            maxQ.push_back(right);
            while (!minQ.empty() && nums[right] <= nums[minQ.back()]) minQ.pop_back();
            minQ.push_back(right);
            
            while (!maxQ.empty() && !minQ.empty() && nums[maxQ.front()] - nums[minQ.front()] > limit) {
                left++;
                while (maxQ.front() < left) maxQ.pop_front();
                while (minQ.front() < left) minQ.pop_front();
            }
            res = max(res, right - left + 1);
        }
        return res;
    }
};

File: rust_solutions/1438.rs
use std::collections::BinaryHeap;
use std::cmp::max;

impl Solution {
    pub fn longest_subarray(nums: Vec<i32>, limit: i32) -> i32 {
        let mut max_q = BinaryHeap::from([(nums[0], 0)]);
        let mut min_q = BinaryHeap::from([(-nums[0], 0)]);

        let mut m = 1;
        let mut left = 0;
        for right in 1..nums.len() {
            max_q.push((nums[right], right));
            min_q.push((-nums[right], right));

            let mut max_t = max_q.peek().unwrap().clone();
            let mut min_t = min_q.peek().unwrap().clone();

            let mut d = max_t.0 + min_t.0;

            while d > limit {
                if min_t.1 < max_t.1 {
                    min_q.pop();

                    if left <= min_t.1 {
                        left = min_t.1 + 1;
                    }
                }
                else {
                    max_q.pop();
                    if left <= max_t.1 {
                        left = max_t.1 + 1;
                    }
                }
                max_t = max_q.peek().unwrap().clone();
                min_t = min_q.peek().unwrap().clone();

                d = max_t.0 + min_t.0;
            }
            m = max(m, right - left + 1);
        }
        return m as i32;
    }
}
